### PR TITLE
Migrate to pyproject.toml

### PR DESCRIPTION
**Pull request by maximlt**
In this PR hvPlot's build infrastructure is migrated to `pyproject.toml`.

- Removed `pyctdev`
- But the dependencies can still be installed with conda. There is now a pre-commit hook that leverages [pyproject2conda](https://github.com/usnistgov/pyproject2conda) to keep a few conda environment files up-to-date with the pyproject.toml file. These conda envs are all used on the CI. The pyproject -> conda conversion offered by `pyproject2conda` isn't yet perfect (I had to add a small post-process script), but that's better than having to maintain the dependencies and their pins manually. I also like that contributors won't be exposed to the tool at all.
- Developer guide updated accordingly (planning to do more work on that front in future PRs)

I've decided not to commit yet to `hatch` (or `pdm`, `poetry`, etc.), this can be done in a second step when the time comes. But this PR will make it a lot easier to transition to such a tool

TODOs:

- [x] pip test suite passes
- [x] conda test suite passes
- [x] docs build with pip
- [x] docs build with conda
- [x] sdist builds, compare with before
- [x] wheel builds, compare with before
- [x] conda package builds, compare with before
- [ ] ~~caching envs on the CI~~ (left for another PR)
- [x] clean up old files

### CODE REVIEW

**Comment by maximlt** [comment]
Tried to re-enable caching of conda envs but got this error below, and anyway it's less important now that the tests mainly run on `pip`. For future reference, except the mentioned error this commit looked in a pretty good state https://github.com/holoviz/hvplot/pull/1296/commits/39e424472534981368a7da919392c1c4b336c612.

```
Run actions/cache/restore@v4
Received 146800640 of 905345882 (16.2%), 140.0 MBs/sec
Received 352321536 of 905345882 (38.9%), 167.3 MBs/sec
Received 570425344 of 905345882 (63.0%), 180.8 MBs/sec
Received 780140544 of 905345882 (86.2%), 185.6 MBs/sec
Cache Size: ~863 MB (905345882 B)
/usr/bin/tar -xf /home/runner/work/_temp/00c48f29-e618-4f4a-a99b-36085e11a125/cache.tzst -P -C /home/runner/work/hvplot/hvplot --use-compress-program unzstd
Received 905345882 of 905345882 (100.0%), 172.4 MBs/sec
/usr/bin/tar: ../../../../../usr/share/miniconda/envs: Cannot utime: Operation not permitted
/usr/bin/tar: ../../../../../usr/share/miniconda/envs: Cannot change mode to rwxr-xr-x: Operation not permitted
/usr/bin/tar: Exiting with failure status due to previous errors
Warning: Failed to restore: "/usr/bin/tar" failed with error: The process '/usr/bin/tar' failed with exit code 2
Cache not found for input keys: test-ubuntu-latest-36c094ec97596d8ea992ee328318c07114d87fdfcacea96199ba303df0212202-2024-03-24
```

Caching of pip envs/packages is left for another PR.

File: doc/developer_guide/index.md [hunk 1]
```diff
@@ -55,49 +52,44 @@ checkout* for the remainder of this document.
...
 
-Since hvPlot interfaces with a large range of different libraries the
-full test suite requires a wide range of dependencies. To make it
-easier to install and run different parts of the test suite across
-different platforms hvPlot uses a library called `pyctdev` to make things
-more consistent and general. Specify also the desired Python version you want
-to base your environment on.
+```
+python -m venv .venv
+source .venv/bin/activate
```
Comment (Hoxbro) [inline doc/developer_guide/index.md:69]: I think this is different on Windows.

File: .flake8 [hunk 2]
```diff
@@ -0,0 +1,7 @@
+[flake8]
```
Comment (Hoxbro) [inline .flake8:1]: Maybe it's time to switch to Ruff?

File: pyproject.toml [hunk 3]
```diff
@@ -1,10 +1,157 @@
...
+    "streamz >=0.3.0",
+    "xarray >=0.18.2",
+]
+tests-nb = [
+    "pytest-xdist",
+    "nbval",
+]
+# Run the example tests by installing examples_tests together with tests
+examples-tests = [
+    "hvplot[examples]",
```
Comment (Hoxbro) [inline pyproject.toml:125]: I am curious: does this work if I change a package in examples?

File: pyproject.toml [hunk 4]
```diff
@@ -1,10 +1,157 @@
...
+    "hvplot[tests-nb]",
+]
+# Additional packages required to build the docs
+doc = [
+    "hvplot[examples]",
+    "nbsite >=0.8.4",
+    "sphinxext-rediraffe",
+]
+# Trick to let pip know we want to install dev dependencies
+hvdev = [
```
Comment (Hoxbro) [inline pyproject.toml:135]: Cool, did not know this trick.

File: update_conda_envs.py [hunk 5]
```diff
@@ -0,0 +1,30 @@
...
+])
+
+# Hacky way to install the package in editable mode when creating the env.
+for file in Path('envs').glob('*.yaml'):
+    with open(file, 'a', encoding='utf-8') as f:
+        f.write('  - pip\n')
+        f.write('  - pip:\n')
+        f.write('    - -e ..\n')
+
+# Remove ibis-duck from the Python 3.12 env since it's python-duck
```
Comment (Hoxbro) [inline update_conda_envs.py:20]: Released over the weekend if you are talking abount python-duckdb

File: update_conda_envs.py [hunk 6]
```diff
@@ -0,0 +1,30 @@
...
+from pathlib import Path
+
+subprocess.run([
+    "pyproject2conda",
+    "project",
+    "--overwrite",
+    "force",
+    "--template-python",
+    str(Path('envs', 'py{py_version}-{env}')),
+])
```
Comment (Hoxbro) [inline update_conda_envs.py:11]: Maybe `check=True`

File: update_conda_envs.py [hunk 7]
```diff
@@ -0,0 +1,30 @@
+import subprocess
```
Comment (Hoxbro) [inline update_conda_envs.py:1]: I like having a `scripts` folder for files like this.

File: doc/developer_guide/index.md [hunk 8]
```diff
@@ -55,49 +52,44 @@ checkout* for the remainder of this document.
...
 
-You will need to pick a Python version. The best practice is to choose the minimum version
-currently supported by hvPlot on the main development branch. If you cannot get the minimum
-version installed, then try with a more recent version of Python.
+Install the test dependencies:
 
-```sh
-conda install python=3.x pyctdev
+``` bash
+pip install -e '.[tests, examples-tests, geo, hvdev, hvdev-geo]'
```
Comment (Hoxbro) [inline doc/developer_guide/index.md:75]: You could split this into two commands, one with and one without geo dependencies. 

Another thing that could be mentioned is that hvdev pulls in dev versions of other holoviz packages.

**Review by Hoxbro — COMMENTED** [review]
I have left some minor comments. But it looks good to me. 

Maybe double-check that the package size matches the size of packages before this PR  (if you haven't already done that).

File: .pre-commit-config.yaml [hunk 9]
```diff
@@ -36,3 +36,12 @@ repos:
       exclude: (\.min\.js$|\.svg$|\.html$)
       additional_dependencies:
       - tomli
+-   repo: local
+    hooks:
```
Comment (Hoxbro) [inline .pre-commit-config.yaml:40]: Will this exit non-zero if changes were made by the script?

File: .flake8 (hunk 2 above)
Comment (maximlt) [reply .flake8:1]: Yes, but not in this PR :)

File: pyproject.toml (hunk 3 above)
Comment (maximlt) [reply pyproject.toml:125]: Yes.

File: update_conda_envs.py (hunk 5 above)
Comment (maximlt) [reply update_conda_envs.py]: Perfect!

File: doc/developer_guide/index.md (hunk 8 above)
Comment (maximlt) [reply doc/developer_guide/index.md:82]: > You could split this into two commands, one with and one without geo dependencies.

I think the developer guide will be refined over time, that's just the first pass, I'm fine with how it is at the moment. In a few months when hvPlot drops support of Python 3.8 hopefully the non-geo/geo stuff can be removed.

> Another thing that could be mentioned is that hvdev pulls in dev versions of other holoviz packages.

Same, I think some more in-depth explanation about the relationship between hvPlot and other packages, HoloViews in particular, is needed for hvPlot contributors.

File: .pre-commit-config.yaml (hunk 9 above)
Comment (maximlt) [reply .pre-commit-config.yaml:40]: pre-commit seems to be smart enough to prevent committing when executing the script leads to changes in the repo.

**Pull request merged**